File: Digital Business Enablement/bluegather/src/main/java/br/com/fiap/bluegather/service/EventoService.java
package br.com.fiap.bluegather.service;

import br.com.fiap.bluegather.dto.EventoDTO;
import br.com.fiap.bluegather.dto.EventoResponse;
import br.com.fiap.bluegather.dto.ImagemDTO;
import br.com.fiap.bluegather.dto.StatusDTO;
import br.com.fiap.bluegather.dto.TipoEventoDTO;
import br.com.fiap.bluegather.dto.UsuarioDTO;
import br.com.fiap.bluegather.model.Evento;
import br.com.fiap.bluegather.model.Imagem;
import br.com.fiap.bluegather.repository.EventoRepository;
import br.com.fiap.bluegather.repository.UsuarioRepository;
import br.com.fiap.bluegather.model.Usuario;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.Collections;
import java.util.Date;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class EventoService {

    @Autowired
    private EventoRepository eventoRepository;

    @Autowired
    private StatusService statusService;

    @Autowired
    private TipoEventoService tipoEventoService;

    @Autowired
    private UsuarioService usuarioService;

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private ImagemService imagemService;

    public Page<EventoDTO> listAll(Pageable pageRequest) {
        Page<Evento> list = eventoRepository.findAll(pageRequest);
        return list.map(this::convertToDto);
    }

    public EventoDTO findById(Long id) {
        Evento entity = findEntityById(id);
        return convertToDto(entity);
    }

    @Transactional
    public EventoDTO create(EventoDTO newData) {
        Evento entity = convertToEntity(newData);
        Evento savedEntity = eventoRepository.save(entity);
        return convertToDto(savedEntity);
    }

    @Transactional
    public EventoDTO update(Long id, EventoDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Evento updatedEntity = convertToEntity(updatedData);    
        Evento savedEntity = eventoRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }
    
    @Transactional
    public void delete(Long id) {
        Evento entity = findEntityById(id);
        eventoRepository.delete(entity);
    }

    public Evento findEntityById(Long id) {
        return eventoRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Evento não encontrado(a) por ID: " + id));
    }

    public Set<EventoResponse> listAllEventoResponses() {
        Set<Evento> eventos = new LinkedHashSet<>(eventoRepository.findAll());
        return eventos.stream()
                      .map(evento -> convertToEventoResponse(evento, imagemService.findByEvento(evento)))
                      .collect(Collectors.toSet());
    }

    public EventoResponse findEventoResponseById(Long id) {
        Evento evento = findEntityById(id);
        Set<Imagem> imagens = imagemService.findByEvento(evento);
        return convertToEventoResponse(evento, imagens);
    }

    public Set<EventoDTO> findByTituloContainingIgnoreCase(String titulo) {
        Set<Evento> list = eventoRepository.findByTituloContainingIgnoreCase(titulo);
        return list.stream().map(this::convertToDto).collect(Collectors.toSet());
    }

    public List<EventoDTO> findByLocalizacaoERaio(double latitude, double longitude, double raioKm) {
        List<Evento> eventos = eventoRepository.findAll();
        List<Evento> eventosFiltrados = eventos.stream()
                .filter(evento -> calcularDistancia(latitude, longitude, Double.parseDouble(evento.getLatitude()), Double.parseDouble(evento.getLongitude())) <= raioKm)
                .collect(Collectors.toList());
        return eventosFiltrados.stream().map(this::convertToDto).collect(Collectors.toList());
    }

    private double calcularDistancia(double lat1, double lon1, double lat2, double lon2) {
        final int R = 6371;
        double latDistance = Math.toRadians(lat2 - lat1);
        double lonDistance = Math.toRadians(lon2 - lon1);
        double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
                * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }

    public Set<EventoDTO> findByDataInicioAfter(LocalDate dataInicio) {
        Date date = Date.from(dataInicio.atStartOfDay(ZoneId.systemDefault()).toInstant());
        Set<Evento> eventos = eventoRepository.findByDataInicioAfter(date);
        return eventos.stream().map(this::convertToDto).collect(Collectors.toSet());
    }

    public Set<EventoDTO> findByUrgencia(Long urgencia) {
        Set<Evento> eventos = eventoRepository.findByUrgencia(urgencia);
        return eventos.stream().map(this::convertToDto).collect(Collectors.toSet());
    }

    public Set<EventoDTO> findByOrganizadorId(Long organizadorId) {
        Set<Evento> eventos = eventoRepository.findByOrganizadorId(organizadorId);
        return eventos.stream().map(this::convertToDto).collect(Collectors.toSet());
    }

    public Set<EventoDTO> findByTipoEventoId(Long tipoEventoId) {
        Set<Evento> eventos = eventoRepository.findByTipoEventoId(tipoEventoId);
        return eventos.stream().map(this::convertToDto).collect(Collectors.toSet());
    }

    public Set<EventoDTO> findByStatusId(Long statusId) {
        Set<Evento> eventos = eventoRepository.findByStatusId(statusId);
        return eventos.stream().map(this::convertToDto).collect(Collectors.toSet());
    }

    public Set<EventoResponse> findEventoResponsesByVoluntarioId(Long voluntarioId) {
        Set<Evento> eventos = eventoRepository.findByVoluntariosId(voluntarioId);
        return eventos.stream().map(evento -> convertToEventoResponse(evento, imagemService.findByEvento(evento))).collect(Collectors.toSet());
    }
    
    public EventoDTO convertToDto(Evento entity) {
        EventoDTO dto = new EventoDTO();
        dto.setId(entity.getId());
        dto.setTitulo(entity.getTitulo());
        dto.setLatitude(entity.getLatitude());
        dto.setLongitude(entity.getLongitude());
        dto.setDataInicio(entity.getDataInicio());
        dto.setDataFim(entity.getDataFim());
        dto.setDescricao(entity.getDescricao());
        dto.setUrgencia(entity.getUrgencia());
        dto.setIdOrganizador(entity.getOrganizador() != null ? entity.getOrganizador().getId() : null);
        dto.setIdTipoEvento(entity.getTipoEvento() != null ? entity.getTipoEvento().getId() : null);
        dto.setIdStatus(entity.getStatus() != null ? entity.getStatus().getId() : null);
        if (entity.getVoluntarios() != null) {
            Set<Long> idsVoluntarios = entity.getVoluntarios().stream()
                    .map(Usuario::getId)
                    .collect(Collectors.toSet());
            dto.setIdsVoluntarios(idsVoluntarios);
        }
        return dto;
    }

    public Evento convertToEntity(EventoDTO dto) {
        if (Objects.isNull(dto)) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - EventoDTO não pode ser nulo.");
        }
        Evento entity;
        if (dto.getId() != null) {
            entity = findEntityById(dto.getId());
        } else {
            entity = new Evento();
        }
        entity.setTitulo(dto.getTitulo());
        entity.setLatitude(dto.getLatitude());
        entity.setLongitude(dto.getLongitude());
        entity.setDataInicio(dto.getDataInicio());
        entity.setDataFim(dto.getDataFim());
        entity.setDescricao(dto.getDescricao());
        entity.setUrgencia(dto.getUrgencia());
    
        if (dto.getIdOrganizador() != null) {
            entity.setOrganizador(usuarioService.findEntityById(dto.getIdOrganizador()));
        } else {
            entity.setOrganizador(null);
        }
    
        if (dto.getIdTipoEvento() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID TipoEvento não pode ser nulo.");
        }
        entity.setTipoEvento(tipoEventoService.findEntityById(dto.getIdTipoEvento()));
    
        if (dto.getIdStatus() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID Status não pode ser nulo.");
        }
        entity.setStatus(statusService.findEntityById(dto.getIdStatus()));
    
        Set<Usuario> newUsuarios = new LinkedHashSet<>();
        if (dto.getIdOrganizador() != null) {
            newUsuarios.add(usuarioService.findEntityById(dto.getIdOrganizador()));
        }
        if (dto.getIdsVoluntarios() != null) {
            dto.getIdsVoluntarios().forEach(id -> {
                Usuario usuario = usuarioRepository.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Usuário não encontrado com ID: " + id));
                newUsuarios.add(usuario);
            });
        }
        entity.setVoluntarios(newUsuarios);
    
        return entity;
    }    

    public EventoResponse convertToEventoResponse(Evento entity, Set<Imagem> imagens) {
        if (entity == null) {
            throw new IllegalArgumentException("Evento não pode ser nulo.");
        }

        Set<ImagemDTO> imagemDTOs = imagens != null ? imagens.stream()
                                                        .map(imagemService::convertToDto)
                                                        .collect(Collectors.toSet()) : Collections.emptySet();
        
        Set<UsuarioDTO> voluntariosDTOs = entity.getVoluntarios() != null ? entity.getVoluntarios().stream()
                                                                        .map(usuarioService::convertToDto)
                                                                        .collect(Collectors.toSet()) : Collections.emptySet();

        UsuarioDTO organizadorDTO = entity.getOrganizador() != null ? usuarioService.convertToDto(entity.getOrganizador()) : null;
        TipoEventoDTO tipoEventoDTO = entity.getTipoEvento() != null ? tipoEventoService.convertToDto(entity.getTipoEvento()) : null;
        StatusDTO statusDTO = entity.getStatus() != null ? statusService.convertToDto(entity.getStatus()) : null;

        return new EventoResponse(
                entity.getId(),
                entity.getTitulo(),
                entity.getLatitude(),
                entity.getLongitude(),
                entity.getDataInicio(),
                entity.getDataFim(),
                entity.getDescricao(),
                entity.getUrgencia(),
                organizadorDTO,
                tipoEventoDTO,
                statusDTO,
                voluntariosDTOs,
                imagemDTOs
        );
    }
}

File: Enterprise Application Development/bluegather/Services/EventoService.cs
using Microsoft.EntityFrameworkCore;
using BlueGather.Dto;
using BlueGather.Models;
using BlueGather.Context;

namespace BlueGather.Services;
public class EventoService
{
    private readonly BlueGatherContext _context;

    public EventoService(BlueGatherContext context)
    {
        _context = context;
    }

    public async Task<List<EventoDto>> FindAll()
    {
        var list = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<EventoDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<EventoDto> Create(EventoDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Evento.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<EventoDto> Update(long id, EventoDto updatedData)
    {
        var entity = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .FirstOrDefaultAsync(x => x.Id == id);

        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}.");
        }

        entity.Voluntarios.Clear();

        if (updatedData.IdsVoluntarios != null)
        {
            var newVoluntarios = await _context.Usuario.Where(v => updatedData.IdsVoluntarios.Contains(v.Id)).ToListAsync();
            foreach (var voluntario in newVoluntarios)
            {
                entity.Voluntarios.Add(voluntario);
            }
        }

        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Evento.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<EventoModel> FindEntityById(long id)
    {
        var entity = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<EventoDto>> FindByUrgencia(long urgencia)
    {
        var list = await _context.Evento
            .Where(x => x.Urgencia == urgencia)
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<EventoDto>> FindByOrganizadorId(long id)
    {
        var list = await _context.Evento
            .Where(x => x.Organizador.Id == id)
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<EventoDto>> FindByTipoEventoId(long id)
    {
        var list = await _context.Evento
            .Where(x => x.TipoEvento.Id == id)
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<EventoDto>> FindByStatusId(long id)
    {
        var list = await _context.Evento
            .Where(x => x.Status.Id == id)
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<EventoDto>> FindByTitulo(string titulo)
    {
        var list = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .Where(x => x.Titulo.ToLower().Contains(titulo.ToLower()))
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<EventoDto>> FindByLocalizacaoERaio(double latitude, double longitude, double raioKm)
    {
        var eventos = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .ToListAsync();

        var eventosFiltrados = eventos
            .Where(evento => CalcularDistancia(latitude, longitude, double.Parse(evento.Latitude), double.Parse(evento.Longitude)) <= raioKm)
            .ToList();

        return eventosFiltrados.Select(ConvertToDto).ToList();
    }

    private double CalcularDistancia(double lat1, double lon1, double lat2, double lon2)
    {
        const int R = 6371;
        var latDistance = ToRadians(lat2 - lat1);
        var lonDistance = ToRadians(lon2 - lon1);
        var a = Math.Sin(latDistance / 2) * Math.Sin(latDistance / 2)
                + Math.Cos(ToRadians(lat1)) * Math.Cos(ToRadians(lat2))
                * Math.Sin(lonDistance / 2) * Math.Sin(lonDistance / 2);
        var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
        return R * c;
    }

    private double ToRadians(double angle)
    {
        return angle * (Math.PI / 180);
    }

    public async Task<HashSet<EventoDto>> FindByDataInicioAfter(DateTime dataInicio)
    {
        var eventos = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .Where(e => e.DataInicio >= dataInicio)
            .ToListAsync();

        return eventos.Select(ConvertToDto).ToHashSet();
    }

    public async Task<HashSet<EventoDto>> FindByVoluntarioId(long id)
    {
        var eventos = await _context.Evento
            .Include(x => x.Organizador)
            .Include(x => x.TipoEvento)
            .Include(x => x.Status)
            .Include(x => x.Voluntarios)
            .Where(e => e.Voluntarios.Any(v => v.Id == id))
            .ToListAsync();

        return eventos.Select(ConvertToDto).ToHashSet();
    }

    private EventoDto ConvertToDto(EventoModel entity)
    {
        return new EventoDto
        {
            Id = entity.Id,
            Titulo = entity.Titulo,
            Latitude = entity.Latitude,
            Longitude = entity.Longitude,
            DataInicio = entity.DataInicio,
            DataFim = entity.DataFim,
            Descricao = entity.Descricao,
            Urgencia = entity.Urgencia,
            IdOrganizador = entity.Organizador?.Id,
            IdTipoEvento = entity.TipoEvento.Id,
            IdStatus = entity.Status.Id,
            IdsVoluntarios = entity.Voluntarios?.Select(x => x.Id).ToList()
        };
    }

    private async Task<EventoModel> ConvertToEntity(EventoDto dto)
    {
        var organizador = await _context.Usuario.FindAsync(dto.IdOrganizador);
        if (organizador == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdOrganizador}");
        }

        var tipo = await _context.TipoEvento.FindAsync(dto.IdTipoEvento);
        if (tipo == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdTipoEvento}");
        }

        var status = await _context.Status.FindAsync(dto.IdStatus);
        if (status == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdStatus}");
        }

        var voluntarios = new List<UsuarioModel>();
        if (dto.IdsVoluntarios != null && dto.IdsVoluntarios.Any())
        {
            voluntarios = await _context.Usuario.Where(voluntario => dto.IdsVoluntarios.Contains(voluntario.Id)).ToListAsync();
        }

        return new EventoModel
        {
            Id = dto.Id ?? 0,
            Titulo = dto.Titulo,
            Latitude = dto.Latitude,
            Longitude = dto.Longitude,
            DataInicio = dto.DataInicio,
            DataFim = dto.DataFim,
            Descricao = dto.Descricao,
            Urgencia = dto.Urgencia,
            Organizador = organizador,
            TipoEvento = tipo,
            Status = status,
            Voluntarios = voluntarios
        };
    }
}